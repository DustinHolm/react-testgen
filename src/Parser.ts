import { existsSync } from "fs"
import { ArrowFunction, ConstructorDeclaration, FunctionDeclaration, FunctionExpression, MethodDeclaration, Project, PropertySignature, SourceFile, SyntaxKind } from "ts-morph"
import { Attribute, ClassElement, Constructor, Export, ExportType, FunctionElement, HasReturnAndParameters, ImportBlock, InterfaceElement, MethodElement, ParserResult, VariableElement } from "./types"
import { createModulePath } from "./utils"


class Parser {
    private source: SourceFile
    private imports: ImportBlock[]
    private isImportsParsed: boolean
    private exports: Export[]
    private isExportsParsed: boolean
    private interfaces: InterfaceElement[]

    constructor(path: string) {
        const project = new Project()
        project.addSourceFileAtPath(path)
        const source = project.getSourceFile(path)
        if (source === undefined) {
            throw new Error("Sourcefile not found!")
        }

        this.source = source
        this.imports = []
        this.isImportsParsed = false
        this.exports = this.getExportNames()
        this.isExportsParsed = false
        this.interfaces = []
    }

    parse(): ParserResult {
        if (!this.isImportsParsed) {
            this.parseImports()
            this.isImportsParsed = true
        }

        if (!this.isExportsParsed) {
            this.parseExports()
            this.isExportsParsed = true
        }

        return {
            imports: this.imports,
            exports: this.exports
        }
    }

    parseExportsOnly(): ParserResult {
        if (!this.isExportsParsed) {
            this.parseExports()
            this.isExportsParsed = true
        }

        return {
            imports: [],
            exports: this.exports
        }
    }

    private parseImports(): void {
        this.imports = this.source.getImportDeclarations().map(imDe => {
            let importModule = imDe.getModuleSpecifier().getText()
            if (importModule.startsWith("\"") && importModule.endsWith("\"")
                || importModule.startsWith("\'") && importModule.endsWith("\'")) {
                importModule = importModule.slice(1, -1)
            }
            const importModuleIsInternal = importModule.startsWith(".")

            const namedImports = imDe.getNamedImports().map(imSp =>
                imSp.getName())
            const defaultImport = imDe.getDefaultImport()?.getText()

            const imports: Export[] = namedImports
                .map(i => ({ name: i, isDefault: false }))
                .concat([defaultImport]
                    .filter(i => i !== undefined)
                    .map(i => ({ name: i!, isDefault: true }))
                )

            if (importModuleIsInternal) {
                const path = createModulePath(this.source.getDirectoryPath(), importModule)
                const fileExtension = existsSync(path + ".tsx") ? ".tsx" : ".ts"
                if (fileExtension === ".ts" && !existsSync(path + fileExtension)) {
                    throw Error(`This import was not found: ${path + fileExtension}`)
                }
                const importParser = new Parser(path + fileExtension)
                const importExports = importParser.parseExportsOnly().exports

                imports.forEach(i => {
                    i.element = importExports.find(e => e.name === i.name)?.element
                })
            }

            return {
                sourceFile: importModule,
                isInternal: importModuleIsInternal,
                imports: imports
            }
        })
    }

    private parseExports(): void {
        this.parseInterfaces()
        this.parseClasses()
        this.parseFunctions()
        this.parseVariables()
    }

    private parseInterfaces(): void {
        this.source.getInterfaces().forEach(inDe => {
            const attributes: Attribute[] = inDe.getMembers().map(tyElTy => ({
                name: (tyElTy as PropertySignature).getName(),
                type: tyElTy.getType().getText()
            }))

            const interfaceElement = {
                returnsJSX: false,
                type: ExportType.Interface,
                attributes: attributes,
                name: inDe.getName()
            } as InterfaceElement

            const existingExport = this.exports.find(e => e.name === inDe.getName())

            if (existingExport !== undefined) {
                existingExport.element = interfaceElement
            }

            this.interfaces.push(interfaceElement)
        })
    }

    private parseClasses(): void {
        this.source.getClasses().forEach(clDe => {
            const existingExport = this.exports.find(e => e.name === clDe.getName())

            if (existingExport !== undefined) {
                const methods = clDe.getMethods().map(meDe => {
                    const name = meDe.getName()
                    const functionElement = this.getFunctionElement(meDe)

                    return {
                        ...functionElement,
                        name: name
                    } as MethodElement
                })

                const returnsJSX = methods.some(m => m.returnsJSX)

                const constructors: Constructor[] = clDe.getConstructors().map(coDe => {
                    const attributes: Attribute[] = this.getAttributes(coDe)

                    return {
                        parameters: attributes
                    }
                })

                existingExport.element = {
                    returnsJSX: returnsJSX,
                    type: ExportType.Class,
                    methods: methods,
                    constructors: constructors
                } as ClassElement
            }
        })
    }

    private parseFunctions(): void {
        this.source.getFunctions().forEach(fuDe => {
            const existingExport = this.exports.find(e => e.name === fuDe.getName())

            if (existingExport !== undefined) {
                existingExport.element = this.getFunctionElement(fuDe)
            }
        })
    }

    private parseVariables(): void {
        this.source.getVariableDeclarations().forEach(vaDe => {
            const existingExport = this.exports.find(e => e.name === vaDe.getName())

            if (existingExport !== undefined) {
                const arrowFunction = vaDe.getInitializerIfKind(SyntaxKind.ArrowFunction)
                const otherFunction = vaDe.getInitializerIfKind(SyntaxKind.FunctionExpression)

                if (arrowFunction) {
                    existingExport.element = this.getFunctionElement(arrowFunction)
                } else if (otherFunction) {
                    existingExport.element = this.getFunctionElement(otherFunction)
                } else {
                    const aliasType = vaDe.getChildrenOfKind(SyntaxKind.TypeReference)
                        .map(typeRef => typeRef.getText())
                        .shift()
                    const variableType = aliasType ?? vaDe.getType().getText()

                    existingExport.element = {
                        returnsJSX: variableType.startsWith("JSX"),
                        type: ExportType.Variable,
                        varType: variableType,
                        varValue: vaDe.getInitializer()?.getText() ?? "undefined"
                    } as VariableElement
                }
            }
        })
    }

    private getExportNames(): Export[] {
        const defaultExports = this.source.getExportAssignments()
        const exportSymbols = this.source.getExportSymbols()
        const exportNames: Export[] = []

        defaultExports.forEach(exAs => {
            exportNames.push({
                name: exAs.getExpression().getText(),
                isDefault: true
            })
        })

        exportSymbols.forEach(exSy => {
            if (exSy.getName() !== "default") {
                exportNames.push({
                    name: exSy.getName(),
                    isDefault: false
                })
            }
        })

        return exportNames
    }

    
    private getFunctionElement(declaration: HasReturnAndParameters): FunctionElement {
        const returnType = declaration.getReturnType().getText()
        const attributes: Attribute[] = this.getAttributes(declaration)

        return {
            returnsJSX: returnType.startsWith("JSX"),
            type: ExportType.Function,
            returnType: returnType,
            parameters: attributes
        }
    }

    private getAttributes(declaration: HasReturnAndParameters | ConstructorDeclaration): Attribute[] {
        return declaration.getParameters().flatMap(p => {
            const typeName = p.getType().getText()
            const correspondingInterface = this.interfaces.find(i =>
                i.name === typeName
                || (typeName.includes("import(")
                    && typeName.includes(i.name)))
            if (correspondingInterface !== undefined) {
                return correspondingInterface.attributes
            } else {
                return {
                    name: p.getName(),
                    type: p.getType().getText()
                }
            }
        })
    }
}

export default Parser