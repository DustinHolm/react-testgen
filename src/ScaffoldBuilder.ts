import { createComponentRenderFunction } from "./templates/componentRenderingTemplate"
import { createDescribeBlock } from "./templates/describeBlockTemplate"
import { createDefaultImports, createImport } from "./templates/importTemplate"
import { createMock } from "./templates/mockTemplate"
import { createProp } from "./templates/propTemplate"
import { createResetMocksFunction } from "./templates/resetMocksTemplate"
import { Export, ImportBlock } from "./types"
import { getPropsForExport } from "./utils"

class ScaffoldBuilder {
    private name: string
    private exports: Export[]
    private imports: ImportBlock[]

    constructor(name: string) {
        this.name = name.replace(".tsx", "").replace(".ts", "")
        this.exports = []
        this.imports = []
    }

    addExport(element: Export) {
        this.exports.push(element)
    }

    addImport(element: ImportBlock) {
        this.imports.push(element)
    }

    build(): string {
        const jsxExports = this.exports.filter(e => e.element?.returnsJSX === true)
        const mainExport: Export | undefined = jsxExports.find(e => e.isDefault) ?? jsxExports[0]
        const props = mainExport !== undefined
            ? getPropsForExport(mainExport)
            : []

        const mocks = this.imports.filter(i => i.isInternal)

        const importsString = [
            createDefaultImports(),
            ...this.imports.map(i => createImport(i))
        ].join("\n")

        const describeString = createDescribeBlock(this.name)

        const resetMocksString = createResetMocksFunction(props, mocks)

        const mocksString = mocks
            .map(i => createMock(i))
            .join("\n")

        const propsString = props
            .map(p => createProp(p))
            .join("\n")

        const componentRenderString = mainExport !== undefined
            ? createComponentRenderFunction(mainExport.name, props)
            : ""

        return `\
${importsString}


${describeString}
${resetMocksString}
${mocksString}
${propsString}
${componentRenderString}
`
    }
}

export default ScaffoldBuilder