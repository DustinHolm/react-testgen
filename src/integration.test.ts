import Parser from "./Parser"
import ScaffoldBuilder from "./ScaffoldBuilder"

describe("Integration Tests", () => {
    test("ReactClass matches snapshot", () => {
        const parser = new Parser("./__test_resources__/ReactClass.tsx")
        const builder = new ScaffoldBuilder("ReactClass.tsx")
        parser.parse().exports.forEach(e => builder.addExport(e))
        parser.parse().imports.forEach(i => builder.addImport(i))
        const returnedString = builder.build()

        expect(returnedString).toMatchSnapshot()
    })

    test("ReactFC matches snapshot", () => {
        const parser = new Parser("./__test_resources__/ReactFC.tsx")
        const builder = new ScaffoldBuilder("ReactFC.tsx")
        parser.parse().exports.forEach(e => builder.addExport(e))
        parser.parse().imports.forEach(i => builder.addImport(i))
        const returnedString = builder.build()

        expect(returnedString).toMatchSnapshot()
    })

    test("ReactFunction matches snapshot", () => {
        const parser = new Parser("./__test_resources__/ReactFunction.tsx")
        const builder = new ScaffoldBuilder("ReactFunction.tsx")
        parser.parse().exports.forEach(e => builder.addExport(e))
        parser.parse().imports.forEach(i => builder.addImport(i))
        const returnedString = builder.build()

        expect(returnedString).toMatchSnapshot()
    })
})