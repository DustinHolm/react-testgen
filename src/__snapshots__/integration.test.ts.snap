// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Tests ReactClass matches snapshot 1`] = `
"import { render, screen } from \\"@testing-library/react\\"
import { Component } from \\"react\\"
import ReactClassDependency from \\"./ReactClassDependency\\"
import { utilitySquareFunction } from \\"./utilityFunctions\\"


describe(\\"ReactClass Tests\\", () => {
    beforeEach(() => {
        resetMocks()
    })

    test(\\"Component renders\\", () => {
        whenComponentIsRendered()
        screen.debug()
    })
})

function resetMocks() {
    resetMockPropContent()
    resetMockReactClassDependency()
    resetMockUtilitySquareFunction()
}

let mockReactClassDependency: jest.Mock
let mockReactClassDependencyReturn: JSX.Element
function resetMockReactClassDependency() {
    mockReactClassDependencyReturn = <span data-testid={\\"mockReactClassDependency\\"} />
    mockReactClassDependency = jest.fn(() => mockReactClassDependencyReturn)
}
jest.mock(\\"./ReactClassDependency\\", () => ({
    __esmodule: true,
    default: (...args: any[]) => mockReactClassDependency(args),
})

let mockUtilitySquareFunction: jest.Mock
let mockUtilitySquareFunctionReturn: number
function resetMockUtilitySquareFunction() {
    mockUtilitySquareFunctionReturn = 1234
    mockUtilitySquareFunction = jest.fn(() => mockUtilitySquareFunctionReturn)
}
jest.mock(\\"./utilityFunctions\\", () => ({
    utilitySquareFunction: (...args: any[]) => mockUtilitySquareFunction(args),
})

let mockPropContent: string
function resetMockPropContent() {
    mockPropContent = \\"abcd\\"
}
function givenMockPropContent(given: string) {
    mockPropContent = given
}

function whenComponentIsRendered() {
    return render(
        <ReactClass
            content={mockPropContent}
        />
    )
}

"
`;

exports[`Integration Tests ReactFC matches snapshot 1`] = `
"import { render, screen } from \\"@testing-library/react\\"
import { FC, useState } from \\"react\\"
import ReactFCDependency from \\"./ReactFCDependency\\"
import { utilitySquareFunction } from \\"./utilityFunctions\\"


describe(\\"ReactFC Tests\\", () => {
    beforeEach(() => {
        resetMocks()
    })

    test(\\"Component renders\\", () => {
        whenComponentIsRendered()
        screen.debug()
    })
})

function resetMocks() {
    resetMockPropContent()
    resetMockReactFCDependency()
    resetMockUtilitySquareFunction()
}

let mockReactFCDependency: jest.Mock
let mockReactFCDependencyReturn: JSX.Element
function resetMockReactFCDependency() {
    mockReactFCDependencyReturn = <span data-testid={\\"mockReactFCDependency\\"} />
    mockReactFCDependency = jest.fn(() => mockReactFCDependencyReturn)
}
jest.mock(\\"./ReactFCDependency\\", () => ({
    __esmodule: true,
    default: (...args: any[]) => mockReactFCDependency(args),
})

let mockUtilitySquareFunction: jest.Mock
let mockUtilitySquareFunctionReturn: number
function resetMockUtilitySquareFunction() {
    mockUtilitySquareFunctionReturn = 1234
    mockUtilitySquareFunction = jest.fn(() => mockUtilitySquareFunctionReturn)
}
jest.mock(\\"./utilityFunctions\\", () => ({
    utilitySquareFunction: (...args: any[]) => mockUtilitySquareFunction(args),
})

let mockPropContent: string
function resetMockPropContent() {
    mockPropContent = \\"abcd\\"
}
function givenMockPropContent(given: string) {
    mockPropContent = given
}

function whenComponentIsRendered() {
    return render(
        <ReactFC
            content={mockPropContent}
        />
    )
}

"
`;

exports[`Integration Tests ReactFunction matches snapshot 1`] = `
"import { render, screen } from \\"@testing-library/react\\"
import { useState } from \\"react\\"
import ReactFCDependency from \\"./ReactFCDependency\\"
import { utilitySquareFunction } from \\"./utilityFunctions\\"


describe(\\"ReactFunction Tests\\", () => {
    beforeEach(() => {
        resetMocks()
    })

    test(\\"Component renders\\", () => {
        whenComponentIsRendered()
        screen.debug()
    })
})

function resetMocks() {
    resetMockPropContent()
    resetMockReactFCDependency()
    resetMockUtilitySquareFunction()
}

let mockReactFCDependency: jest.Mock
let mockReactFCDependencyReturn: JSX.Element
function resetMockReactFCDependency() {
    mockReactFCDependencyReturn = <span data-testid={\\"mockReactFCDependency\\"} />
    mockReactFCDependency = jest.fn(() => mockReactFCDependencyReturn)
}
jest.mock(\\"./ReactFCDependency\\", () => ({
    __esmodule: true,
    default: (...args: any[]) => mockReactFCDependency(args),
})

let mockUtilitySquareFunction: jest.Mock
let mockUtilitySquareFunctionReturn: number
function resetMockUtilitySquareFunction() {
    mockUtilitySquareFunctionReturn = 1234
    mockUtilitySquareFunction = jest.fn(() => mockUtilitySquareFunctionReturn)
}
jest.mock(\\"./utilityFunctions\\", () => ({
    utilitySquareFunction: (...args: any[]) => mockUtilitySquareFunction(args),
})

let mockPropContent: string
function resetMockPropContent() {
    mockPropContent = \\"abcd\\"
}
function givenMockPropContent(given: string) {
    mockPropContent = given
}

function whenComponentIsRendered() {
    return render(
        <ReactFunction
            content={mockPropContent}
        />
    )
}

"
`;
